src/main/kotlin/com/openlattice/data/storage/IndexingMetadataManager.kt

```diff
@@ -1,34 +1,83 @@
 package com.openlattice.data.storage
 
 import com.openlattice.data.EntityDataKey
+import com.openlattice.data.storage.partitions.PartitionManager
 import com.openlattice.postgres.DataTables.LAST_INDEX
 import com.openlattice.postgres.DataTables.LAST_LINK
 import com.openlattice.postgres.PostgresArrays
 import com.openlattice.postgres.PostgresColumn.*
 import com.openlattice.postgres.PostgresTable.IDS
 import com.zaxxer.hikari.HikariDataSource
-import org.slf4j.LoggerFactory
+import java.sql.Connection
 import java.time.OffsetDateTime
-import java.util.UUID
-import java.util.Optional
+import java.util.*
 
 
-class IndexingMetadataManager(private val hds: HikariDataSource) {
-    companion object {
-        private val logger = LoggerFactory.getLogger(IndexingMetadataManager::class.java)
+class IndexingMetadataManager(private val hds: HikariDataSource, private val partitionManager: PartitionManager) {
+
+    fun markAsIndexed(entityKeyIdsWithLastWrite: Map<UUID, Map<UUID, OffsetDateTime>>): Int {
+        var count = 0
+
+        hds.connection.use { connection ->
+            entityKeyIdsWithLastWrite.forEach { (entitySetId, entities) ->
+                val partitions = partitionManager.getEntitySetPartitionsInfo(entitySetId).partitions.toList()
+
+                entities.entries
+                        .groupBy({ getPartition(it.key, partitions) }, { it.toPair() })
+                        .mapValues { it.value.toMap() }
+                        .forEach { (partition, entitiesWithLastWrite) ->
+
+                            entitiesWithLastWrite.entries
+                                    .groupBy { it.value }
+                                    .mapValues { it.value.map { it.key } }
+                                    .forEach { (lastWrite, entities) ->
+                                        count += markAsIndexed(connection, entitySetId, entities, partition, lastWrite)
+                                    }
+                        }
+            }
+        }
+        return count
     }
 
-    fun markAsIndexed(entityKeyIdsWithLastWrite: Map<UUID, Map<UUID, OffsetDateTime>>, linking: Boolean): Int {
+    private fun markAsIndexed(
+            connection: Connection,
+            entitySetId: UUID,
+            entityKeyIds: List<UUID>,
+            partition: Int,
+            lastWrite: OffsetDateTime
```

Comment:
Just for clarity it would be good to rename this `lastIndex`

Reply:
Yeah, you're right!